src/graphics/render-pass.js

```diff
@@ -0,0 +1,226 @@
+     */
+    clearStencilValue = 0;
+
+    /**
+     * True if the depth attachment should be cleared before rendering, false to preserve
+     * the existing content.
+     */
+    clearDepth = false;
+
+    /**
+     * True if the stencil  attachment should be cleared before rendering, false to preserve
+     * the existing content.
+     */
+    clearStencil = false;
+
+    /**
+     * True if the depth attachment needs to be stored after the render pass. False
+     * if it can be discarded.
+     *
+     * @type {boolean}
+     */
+    storeDepth = false;
+
+    /**
+     * True if the stencil attachment needs to be stored after the render pass. False
+     * if it can be discarded.
+     *
+     * @type {boolean}
+     */
+    storeStencil = false;
+}
+
+/**
+ * A render pass represents a node in the frame graph, and encapsulates a system which
+ * renders to a render target using an execution callback.
+ *
+ * @ignore
+ */
+class RenderPass {
+    /** @type {string} */
+    name;
+
+    /** @type {RenderTarget} */
+    renderTarget;
+
+    /**
+     * Number of samples. 0 if no render target, otherwise number of samples from the render target,
+     * or the main framebuffer if render target is null.
+     *
+     * @type {number}
+     */
+    samples = 0;
+
+    /** @type {ColorAttachmentOps} */
+    colorOps;
+
+    /** @type {DepthStencilAttachmentOps} */
+    depthStencilOps;
+
+    /**
+     * If true, this pass might use dynamically rendered cubemaps. Use for a case where rendering to cubemap
+     * faces is interleaved with rendering to shadows, to avoid generating cubemap mipmaps. This will likely
+     * be retired modified when render target dependency tracking gets implemented.
+     *
+     * @type {boolean}
+     */
+    requiresCubemaps = true;
+
+    /**
+     * True if the render pass uses full viewport / scissor for the rendering into the render target.
+     *
+     * @type {boolean}
+     */
+    fullSizeClearRect = true;
+
+    /**
+     * Creates an instance of the RenderPass.
+     *
+     * @param {GraphicsDevice} graphicsDevice - The graphics device.
+     * @param {Function} execute - Custom function that is called when the pass needs to be
+     * rendered.
+     */
+    constructor(graphicsDevice, execute) {
+        this.device = graphicsDevice;
+
+        /** @type {Function} */
+        this.execute = execute;
+    }
+
+    /**
+     * @param {RenderTarget} renderTarget - The render target to render into (output). This function should be
+     * called only for render passes which use render target, or passes which render directly into the default
+     * framebuffer, in which case a null or undefined render target is expected.
+     */
+    init(renderTarget) {
+
+        // null represents the default framebuffer
+        this.renderTarget = renderTarget || null;
+
+        // allocate ops only when render target is used
+        if (this.renderTarget !== undefined) {
```

Comment:
```suggestion
        if (this.renderTarget !== null) {
```

Reply:
this would change behavior, this is not desirable

Reply:
the line above means it'll never be undefined

Reply:
did you mean `renderTarget !== undefined` perhaps?

Reply:
true!